### PR TITLE
Schema Support

### PR DESCRIPTION
**Pull request by chandnisingh**

### CODE REVIEW

File: engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlan.java [hunk 1]
```diff
@@ -1179,6 +1186,14 @@ public void validate() throws ConstraintViolationException
           if (pm.portAnnotation != null && !pm.portAnnotation.optional()) {
             throw new ValidationException("Output port connection required: " + n.name + "." + pm.getPortName());
           }
+        } else {
+          //port is connected
+          if (pm.portAnnotation != null && pm.portAnnotation.schemaRequired()) {
+            //since schema is required, the port attribute tuple_class_name should be present
+            if (pm.attributes.get(PortContext.TUPLE_CLASS) == null) {
+              throw new ValidationException("Attribute tuple class name missing on port : " + n.name + "." + pm.getPortName());
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlan.java]: Please use the attribute name in the message.

File: engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlan.java [hunk 2]
```diff
@@ -1170,6 +1170,13 @@ public void validate() throws ConstraintViolationException
               validateThreadLocal(n);
             }
           }
+
+          if (pm.portAnnotation != null && pm.portAnnotation.schemaRequired()) {
+            //since schema is required, the port attribute TUPLE_CLASS should be present
+            if (pm.attributes.get(PortContext.TUPLE_CLASS) == null) {
+              throw new ValidationException("Attribute tuple class missing on port : " + n.name + "." + pm.getPortName());
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlan.java]: Use attribute name.

File: engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlan.java [hunk 3]
```diff
@@ -1179,6 +1186,14 @@ public void validate() throws ConstraintViolationException
           if (pm.portAnnotation != null && !pm.portAnnotation.optional()) {
             throw new ValidationException("Output port connection required: " + n.name + "." + pm.getPortName());
           }
+        } else {
+          //port is connected
+          if (pm.portAnnotation != null && pm.portAnnotation.schemaRequired()) {
+            //since schema is required, the port attribute TUPLE_CLASS should be present
+            if (pm.attributes.get(PortContext.TUPLE_CLASS) == null) {
+              throw new ValidationException("Attribute tuple class missing on port : " + n.name + "." + pm.getPortName());
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlan.java]: Use attribute name.

File: engine/src/main/java/com/datatorrent/stram/webapp/OperatorDiscoverer.java [hunk 4]
```diff
@@ -60,6 +61,8 @@
  */
 public class OperatorDiscoverer
 {
+  public static final String GENERATED_CLASSES_JAR = "dt-generated-classes.jar";
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/webapp/OperatorDiscoverer.java]: Get rid of dt- prefix. How about using prefix '.' or '_' instead?

File: engine/src/main/java/com/datatorrent/stram/webapp/TypeGraph.java [hunk 5]
```diff
@@ -1152,4 +1152,31 @@ public TypeGraph read(Kryo kryo, Input input, Class<TypeGraph> type)
     return result;
   }
 
+  /**
+   * A utility method that tells whether a class is considered a POJO
+   *
+   * @param className name of the class
+   * @return true if it is a POJO false otherwise.
+   */
+  public boolean isInstantiableClassPOJO(String className) throws JSONException
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/webapp/TypeGraph.java]: What is a "classPOJO"? ;-)

Comment (chandnisingh) [inline engine/src/main/java/com/datatorrent/stram/webapp/TypeGraph.java]: Will change this to isInstantiablePojo

File: engine/src/main/java/com/datatorrent/stram/webapp/OperatorDiscoverer.java (hunk 4 above)
Comment (chandnisingh) [inline engine/src/main/java/com/datatorrent/stram/webapp/OperatorDiscoverer.java]: I will change it to use '_'

**Pull request merged**